VirtoCommerce.CatalogPersonalizationModule.Data/Services/PersonalizationService.cs

```diff
@@ -7,117 +8,194 @@
+                }
+
+                if (!string.IsNullOrWhiteSpace(criteria.EntityType))
+                {
+                    query = query.Where(x => x.ObjectType == criteria.EntityType);
+                }
+
+                if (!criteria.Ids.IsNullOrEmpty())
+                {
+                    query = query.Where(x => criteria.Ids.Contains(x.Id));
+                }
+
+                if (criteria.ChangedFrom.HasValue)
+                {
+                    query = query.Where(x => x.ModifiedDate.HasValue && x.ModifiedDate.Value.Date >= criteria.ChangedFrom.Value.Date);
+                }
+
+                var sortInfos = criteria.SortInfos;
+                if (sortInfos.IsNullOrEmpty())
+                {
+                    sortInfos = new[] { new SortInfo { SortColumn = ReflectionUtility.GetPropertyName<TaggedItem>(x => x.Label) } };
+                }
+                query = query.OrderBySortInfos(sortInfos);
+
+                retVal.TotalCount = query.Count();
+                query = query.Skip(criteria.Skip).Take(criteria.Take);
+
+                var ids = query.Select(x => x.Id).ToArray();
+                retVal.Results = GetTaggedItemsByIds(ids, criteria.ResponseGroup).AsQueryable().OrderBySortInfos(sortInfos).ToList();
+
+                if (criteria.ResponseGroup.Contains(TaggedItemResponseGroup.WithInheritedTags.ToString()) &&
+                    !criteria.EntityIds.IsNullOrEmpty())
+                {
+                    var taggedItems = FillInheritedTags(criteria.EntityIds, retVal.Results.ToList());
+                    retVal.TotalCount = taggedItems.Count;
+                    retVal.Results = taggedItems;
+                }
+            }
+
+            return retVal;
+        }
+
+        public void DeleteTaggedItems(string[] ids)
+        {
+            using (var repository = _repositoryFactory())
+            {
+                repository.DeleteTaggedItems(ids);
+                CommitChanges(repository);
+            }
+        }
+
+
+        private List<TaggedItem> FillInheritedTags(string[] entityIds, List<TaggedItem> taggedItems = null)
+        {
+            var result = taggedItems ?? new List<TaggedItem>();
+            var evaluatedEntityIds = new List<string>();
+
+            if (!result.IsNullOrEmpty())
+            {
+                evaluatedEntityIds = result.Select(x => x.EntityId).ToList();
+                foreach (var taggedItem in result)
+                {
+                    var entities = _taggedEntitiesServiceFactory.Create(taggedItem.EntityType).GetEntitiesByIds(new[] { taggedItem.EntityId }).ToList();
+                    var evaluatedItems = EvaluateEffectiveTags(entities);
+                    taggedItem.InheritedTags = evaluatedItems.FirstOrDefault(x => x.EntityId == taggedItem.EntityId)?.InheritedTags;
+                }
+            }
+
+            entityIds = entityIds.Where(ids => evaluatedEntityIds.All(evaluatedIds => evaluatedIds != ids)).ToArray();
```

Comment:
What this code is doing? Seems it is not get all ids except evaluatedIds.
I think there is linq [Except](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.except?view=netframework-4.8) function for this.